### ISSUE TITLE
Run NUTS on baseline model

### ISSUE THREAD

**Issue by EiffL** [DifferentiableUniverseInitiative/jax-cosmo-paper#7]
@jecampagne  we can use this thread to document your implementation of NUTS for the baseline model already used by Joe.

**Comment by EiffL** [comment]
Hi @jecampagne I played a bit with your numpyro code and comparison to cobaya results. 

A few things I took away:
- Your model matches Joe's correctly without problem, and no need to fine tune zmax once [this fix](https://github.com/DifferentiableUniverseInitiative/jax_cosmo/pull/78) is applied. 

- I added some reparameterisation of the problem to try to standardize all variables to more or less the same scales. Specifically I use a `LocScaleReparam` and `TransformReparam` (see[ here](https://colab.research.google.com/github/pyro-ppl/numpyro/blob/master/notebooks/source/bad_posterior_geometry.ipynb) for context). This should prevent the step size from being widely inadapted to some variables vs others

- I authorized full mass matrix estimation `dense_mass=True`, which in principle should help a bit at combating annyoing posterior geometry, but there is a tradeoff because estimating the full mass matrix is difficult, so not sure whether it helps or not in the end.

- Much more importantly, I use parallel sampling with `num_chains=32` and `chain_method='vectorized'` this , will let 32 chains run in parallel on the same GPU. There is a tradeoff there as well, because increasing the number of parallel chains accelerates the total number of samples you can get, up to a point where computing too many chains in parallel slows the evaluation on the GPU. Finding the best balance will depend on the GPU.

All in all, it seems to work decently compared to @joezuntz 's Cobaya run:
![image](https://user-images.githubusercontent.com/861591/139597357-3fbc9af5-932d-4c7e-b3c2-22cddbc13e22.png)
(purple is 15000 samples obtained in ~3h on a single GPU, blue is 147595 cobaya samples obtained on multiple nodes (dont remember exactly how many) )

But the sampling can definitely be optimized, right now it's really not satiftactory, I use max_tree_depth of 5 so that's like maximum 31 steps at each nuts step, and almost always reaches the max. The r_hat and ESS values are not great, the relative effective sample size remains very close to Cobaya's, which is a shame.
For instance:
ESS (Cobya) for sigma8: 0.02919
ESS (NUTS) for sigma8: 0.03064

:-| 

In any case, I've put my demo notebook [here](https://github.com/DifferentiableUniverseInitiative/jax-cosmo-paper/blob/master/notebooks/VectorizedNumPyro.ipynb)

**Comment by EiffL** [comment]
Even if we don't get the sampling to work much more efficiently, there is still something to be said about parallel chains on GPUs :-) My personal GPU is probably twice slower than a A100 such as the ones we will have on Perlmutter. 

So we can roughly expect to achieve 64 x 500 samples per GPU in 3 hours, and you have 4 GPUs per node, so at a bare minimum, we would be looking at about 40,000 samples per node-hour in this configuration.

**Comment by jecampagne** [comment]
Well, I do not know what to say, in fact @EiffL  has answered all the questions I had to @joezuntz. And I do not see any thing to do beyond your nb now.

**Comment by EiffL** [comment]
ahh not quite ^^ I think my NUTS sampling is terribly inefficient, and I'm not sure how to improve that

**Comment by EiffL** [comment]
I get more or less the same sampling efficienct as cobaya:
```
For instance:
ESS (Cobya) for sigma8: 0.02919
ESS (NUTS) for sigma8: 0.03064
```

So this indicates that the gradients are not being very helpful, I'm sure this can be optimized.

**Comment by jecampagne** [comment]
@EiffL 
I am not sure that I have understood your point on zmax. Your [fix](https://github.com/DifferentiableUniverseInitiative/jax_cosmo/pull/78) is definitely what we had discussed about, ok. But, 
here are the kde_plots for the sources
![image](https://user-images.githubusercontent.com/20539759/139861802-c0267741-30b7-4b53-9c4d-d5da21207849.png)
and those for the lenses
![image](https://user-images.githubusercontent.com/20539759/139861925-3c205924-121f-45f5-adfd-d8f72767412f.png)

It is only if I set zmax according to 
```python
nzs_s = [jc.redshift.kde_nz(nz_source['Z_MID'].astype('float32'),
                            nz_source['BIN%d'%i].astype('float32'), 
                            bw=0.01,
                            gals_per_arcmin2=neff_s[i-1], zmax=2.0)
           for i in range(1,5)]
nzs_l = [jc.redshift.kde_nz(nz_lens['Z_MID'].astype('float32'),
                            nz_lens['BIN%d'%i].astype('float32'), bw=0.01, zmax=1)
        for i in range(1,6)]
```
that I get correct matching.
![image](https://user-images.githubusercontent.com/20539759/139862267-bffce123-c694-4fc5-bcc0-8f8bf55b69d5.png)

![image](https://user-images.githubusercontent.com/20539759/139862939-1afa6633-1861-4a77-bd0c-10e99532a0d9.png)


Now, if one compares to Joe's plot, I guess if the he was using kde_plots wo the zmax set then it is why we get the same final plots wo zmax set. Now, if it is correct or no is an other question.

**Comment by jecampagne** [comment]
I just try to run on V100 @ CC in interactif with progress_bar=True, it foresees 28h to run :( 
But V100 and TitanXP are more or less equivalent, so I guess the design of multi-user GPU @ CC limits the running efficiency.

**Comment by jecampagne** [comment]
Concerning the reparametrisation of all Uniform priors to a [-3,3]^d hypercube, why not but Numpyro developper Du Phan told me that Uniform prior is changed into unconstrained prior thanks to sigmoid @ affine transformation.

**Comment by EiffL** [comment]
Ok yeah, but the remapping was to try to make it so all variables had the same order of magnitude. Because it makes HMC easier

Regarding redshift, I don't know it's weird, as I said, my n(z) looks the same whether or not I specify a zmax. Are you using the master branch of Jax cosmo?

**Comment by jecampagne** [comment]
I had launched my script (wo reparametrisation & vectorisation, and with f_sky=0.032) on a GPU batch queue @ CCIn2p3. This is a single chain of 8,000 spls and 1,000 warm-up. Here is the cosmological parameters corner view.
![image](https://user-images.githubusercontent.com/20539759/140298788-a9a228ad-7e75-4e1c-ac7a-018699257f65.png)

As Francois point out, the sampling efficiency is rather low: for sigma8 : 50 eff. spls for a total 8_000 spls generated. Compared to the WL-only case were I got 3 times more eff. spls. 
Also , the bias are highly correlated:
![image](https://user-images.githubusercontent.com/20539759/140299314-9509c7e0-0689-4a2e-8b23-c4491836b8f3.png)

Which suggest a possible difficulty for the sampler to manage that posterior shaping. But we need more investigation.
